Rewrite how NPCs evaluate threats.

I-am-Erk opened the pull request:
<!-- HOW TO USE: Under each "#### Heading" below, enter information relevant to your pull request.
Leave the headings unless they don't apply to your PR.

Please read carefully and don't delete the comments delimited by "< !--" and "-- >"
Once a pull request is submitted, automatic stylistic and consistency checks will be performed on the PR's changes.
The results of these can be either seen under the "Files changed" section of a PR or in the check's details.

NOTE: Please grant permission for repository maintainers to edit your PR.  It is EXTREMELY common for PRs to be held up due to trivial changes being requested and the author being unavailable to make them. -->

#### Summary
Features "Accidentally winds up rewriting how NPCs assess danger."
<!-- This section should consist of exactly one line, edit the one above.
1. Replace the word "Category" with one of these words: Features, Content, Interface, Mods, Balance, Bugfixes, Performance, Infrastructure, Build, I18N.
2. Replace the text inside the quotes with a brief description of your changes.
Or if you don't want a changelog entry, replace the whole line with just the word "None" (with no quotes).
For more on the meaning of each category, see:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/CHANGELOG_GUIDELINES.md
If approved and merged, your summary will be added to the project changelog:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/changelog.txt -->

#### Purpose of change

So, um.

This all started when I decided to refactor `npc::assess_danger()` and add some comments to it. I started noting that NPCs' assessments of how tough their friends were weird:
1. They didn't make sense. A weak ally was getting scores of like 100, saying they could take on a hulk with their fists.
2. They were rapidly changing in a large fight, in what seemed unpredictable patterns, bouncing between 30 and 120 in about five turns at one point.

I finally decided to track down where these numbers came from, and here we are.

Before reading on, be aware that I have the utmost respect for the people who originally coded this. They're two of the people whose code I go to when I want to understand how to do something right. However, well, pobody's nerfect.

The old method for estimating how tough a non-monster target is... it, uh, had a few flaws.
- If the target had a gun, and the assessor didn't, the target was considered more threatening, even if they were an ally. That meant that an NPC swapping from a gun to a sword would then rate its team as having become much stronger.
- Assessment of the target would register it as *more* dangerous the more HP the *assessor* has, and also the more max HP they had on their torso. Not the current HP, the max HP. So as they got hurt they rated themselves, and their enemies, lower. If they got a mutation that increased their max HP, they rated their enemies higher.
- the assessment was also *less* dangerous the stronger the assessor's weapon was. Note that this is also how the assessor rated themselves and their allies, so if they had a stronger weapon they rated themselves and their friends lower. Since they'd also be adding their own weapon into this, the math got complex but basically it meant that the lower their HP got, the more they weighted the value of their weapon. This is a big part of why I saw the numbers jumping all over the place.

While I was at this, I noticed that monsters HP and distance weren't factoring into their difficulty in most regards either, and I rewrote a lot of the calculations there

Basically, I am not sure how NPCs managed to even resemble functionality with this. It's utterly nonsensical, I think the bit about HP is a combination of both incorrect logic and some typos in the implementation of that incorrect logic in the first place. Fixing it is probably going to make a lot of differences in how NPCs assess danger, and also allow us to use those danger assessments in new ways now that they make sense.


#### Describe the solution
1. Refactor the `npc::evaluate_enemy()` and `character_danger()` functions into three functions: `evaluate_monster()`, `evaluate_character()`, and `evaluate_self()`. OK, I didn't so much refactor `character_danger()` as nuke it from orbit and then pick up the pieces and superglue them together into something new.
- `evaluate_monster()` is, for now, basically the same as `evaluate_enemy` used to be, but it only works for monsters. I have some ideas for how to adjust this later, but first let's get this whole deal out.
- `evaluate_character()` can be used to evaluate friends, neutrals, and enemies; other NPCs, or the player character.
- `evaluate_self()` only assesses yourself.
2. Create `npc::estimate_armour()`, which calculates a weighted average of bullet, blunt, cut, and stab armour ratings across all body parts on the target Character, with the torso and head counting 3x as much as other parts. This function should be reasonably futureproofed against adding new body parts.
3. Revise threat calculations. Here is the core threat calculation I started with:
- `threat = ( dodge + armour + weapon value ) * speed * %health remaining`
- When evaluating others, the NPC's perception "fuzzes" their result. The higher the perception, the more accurate it is.  Personality also figures in here.
- If an NPC is running away, their threat is dropped to 50% of its total.
- When evaluating themself, the NPC's own stats are affected by internal stats that they're unaware of on other characters, such as their current level of pain. The amount these affect them is randomized and reduced by perception: a more perceptive NPC is less likely to get overwhelmed by the pain and will keep a more accurate assessment of their wounds.
- a character's bleeding figures into their health assessment. When assessing anyone other than themselves, the NPC assessing can only spot intense bleeding, influenced by their perception score.
- Speed assessment varies slightly depending on if the NPC and/or the target have a gun. Speed matters a bit less for targets wielding guns and a bit less if NPC is wielding a gun.
- Aggression and bravery personality scores also influence assessments, and later I'd like NPC opinions to factor in as well, so if they trust you they'll score you higher.
4. Add the CombatAI debug message category and adds commentary in this category and in the ItemAI category I created in #69420. It's through these categories that I realized this function was broken, and I truly cannot over-emphasize how important it is that we improve logic reporting so that these numbers can be seen and noticed when they don't make sense. This has sat, totally and utterly broken, for seven years because it was quietly happening in a spot where the calculations were too complex to see that they were nonsensical without seeing the numbers. Instead, NPCs have just seemed unpredictable and strange.
- My plan here is to use the NPC_whateverAI message categories for the "nitty gritty" detailed reports, and save the basic DF_NPC reports for just basic updates on the most important stuff. So if you see an NPC making weird combat decisions you can break the details down by turning on CombatAI reports.
5. Adjust how monster difficulty is rated.
- Monsters no longer have a hard lower cap on how low their threat can be. The cap does figure in, but it figures in *before* calculation of distance and HP are added, so a monster's overall threat floors at 5, but if it's far away and injured the NPC can rate it lower than that.
- Distance reduces threat as an inverse square, starting around 10 tiles away (adjusted by monster speed). So after ~10 tiles away a human-speed monster starts getting exponentially less threatening. I need to review the math on this and make sure it's sound, but early tests look good.
- TO DO: Needs some adjustment for monsters that have not yet noticed the assessing NPC vs. monsters that are on alert. Won't happen this PR because I'd like to add it along with some logic about what to do about un-alert monsters.
6. This is a lateral change, but I changed how the `assessment` score works when assessing danger. This used to add in all the enemies and subtract the allies, then compare the result to the assessing NPC's difficulty. Now, I add up all the enemies and add up all the allies, and include the NPC as an ally. The math is the same, I just found it easier to error check this way and it let me remove some 'do not hit zero' checks because there's no subtraction now.

Still TO DO besides that: I would like to have the NPC rate their own weapon value lower if their weapon uses ammo, based on how much ammo they have left and how much they have in inventory. I may not add that in this PR as I have already got 220 lines of diff.

#### Describe alternatives you've considered
I considered some more robust ways to calculate armour and damage, and may still implement these, but for now I think this is working OK and I have other things I want to get done.

I'd still like to add things like the target's strength and weapon skill to its assessed threat, influenced by the NPC's own skill with that weapon, but that's a little ways off and might be too much for this PR. If I get there, I'm worried I'll wind up looking at `weapon_value()` and I'm afraid of what I might see. I'm supposed to be adding memory to NPCs, not rewriting their entire basic logic stuff.

Finally, I really think we should have some CI testing here but I don't know how to add that.

#### Testing

Playtest resutls: It's very hard to do enough play testing to get a real feel for this since it's going to vary wildly with different players, loadouts, and scenarios, but the initial tests are veyr promising
- npc fleeing behaviour seems to be much more predictable.
- My npc with a fire axe did a really good job of ditching combat when the zombie threat got too great
- He actually ran away before the point of no return death stuff, which was great
- then he activated my "don't stray too far from player" caveat, came back, saw I was nearby, bolstered his morale, and ran in and died at my side, which was pretty stupid. I have fixes for this planned next.

Weapon assessments feel OK, the role of armour makes a big difference now, and NPCs notice various important things that are helpful to them. I didn't do a lot of ranged weapon testing except to ensure they were valued more appropriately. I don't want to get in the weeds there, there's a lot of work still to be done with rnaged (but at least they can reload their magazines now, nice)

I could use some help testing against more than just crowds of zombies, and testing things like bandits. I will try to get some more work done on those, but the issue with AI changes is that they are so situational that it's hard to guess what things could go wrong.

---
![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/45136638/affd9a4d-4a7c-4754-b17e-094f3ec97547)
A slightly outdated shot showing the more verbose combat reports. In this example, the NPC has a gun, and is far away from threats so they're rated very low... and you can see that logic happening, what a concept.

---

Final or semifinal testing:

![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/45136638/fa349914-c4b3-4075-925b-961beaa7575a)
Monster threat assessment report

![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/45136638/1ebed054-7a12-49e7-a76c-8c5c18a63008)
Armour and dodge assessment report

![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/45136638/2736ba3e-5b61-419d-a34b-7875b2c6debc)
Further fuzz of assessment based on NPC perception, I'm quite pleased about this one.

![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/45136638/c6985d0f-074b-4de9-9d7b-3aea4bfb6314)
![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/45136638/cddb4b55-5ef3-4203-843a-25644db1adb9)
![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/45136638/daf9f1c5-8f3a-4c8b-ae52-3e988a835d01)
![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/45136638/b953debb-ef13-4d70-9bc1-be9670ecb92c)
![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/45136638/a7699455-1169-424b-a4fa-75abe20a1f9e)
Some examples of the new verbose combat reporting that makes this much easier to debug and assess, and basically opens up the black box of the NPC brain to prying eyes.

![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/45136638/3357e243-d57c-484f-bdef-7a3886cd02cf)
This is another NPC in the exact same combat, she's got lower bravery and is more hurt so she's deciding to run away, frankly it's the better solution. Especially since the player has been standing a little ways away just watching their friends die for several turns and taking notes on a clipboard.

#### Additional context
In this process I learned that melee and ranged weapons are valued on a totally different scale, one where a slingshot counts as a score of 11, and a tempered nodachi counts as a score of 8. I added a quick, very dirty scalar fix to weapon_value() to get the numbers into a comparative scale that at least keeps NPCs working a bit, but this is going to need a much closer look, one I have no capacity for within this PR.

I-am-Erk committed:
refactor and fix evaluate_enemy

I-am-Erk committed:
Cleanup, simplify until maths working

I-am-Erk committed:
add estimate_armour function

refactor/rewrite... nearly... complete

I-am-Erk committed:
add perception and pain effect

I-am-Erk committed:
asstile

I-am-Erk committed:
Update npcmove.cpp

I-am-Erk committed:
Add some more debug messages for testing

I-am-Erk committed:
simplify calc by tallying allies vs enemies

Uncover still more issues in the process

fungamer2-2 commented on src/npcmove.cpp:
```diff
@@ -399,20 +404,141 @@ std::vector<sphere> npc::find_dangerous_explosives() const
+    // the worse pain the NPC is in, the more likely they are to overestimate the severity of their injuries.
+    // the more perceptive they are, the more they're able to see how bad it really is.
+    float pain_factor = rng( 0.0f,
+                             static_cast<float>( get_pain() ) / static_cast<float>( get_per() ) );
```
```suggestion
                             static_cast<float>( get_pain() ) / std::max( 1.0, static_cast<float>( get_per() ) ) );
```

You may want to add a `std::max` here to avoid a potential division by zero.

I-am-Erk replied:
oh yes, I noticed that and then forgot, thanks for reminding me.

I-am-Erk replied:
oh wait it looks like I did do that. Thakns anwyyay

I-am-Erk committed:
this might work now

I-am-Erk committed:
logic and formatting

1. add colour to messages for easier reading
2. adjust error in perception calculation
3. running away NPCs are valued lower.

I-am-Erk committed:
Colourize debug log

because why not

I-am-Erk committed:
fix another math error

monster HP was backwards.

Also updates more colours.

I-am-Erk committed:
math errors math errors aaaah

melee and ranged weapons were not in even vague parity on evaluation.

Did not fix that, but did make it tolerable for now.

I-am-Erk committed:
scale ally threat same as enemy threat

missed a scalar

I-am-Erk committed:
Add a bleeding assessment to threat.

Maleclypse commented:
Exciting stuff!  Oh I think the comment from yesterday is probably only a few minutes ago?   Or is that the 20th?

I-am-Erk commented:
Thanks! I'm excited, now I can push some more visible changes
